## Simulació 

In [1]:
import random
import numpy as np
import pandas as pd
import csv
pd.options.mode.chained_assignment = None  # default='warn'
np.random.seed(6)
# pct = percentage

In [2]:
df = pd.read_csv("Dades/df_train_complete.csv")
lst = ['winner_name','winner_first_serves_in', 'winner_first_serves_total',
         'winner_first_serve_points_won', 'winner_first_serve_points_total',
         'winner_second_serve_points_won', 'winner_second_serve_points_total',
       'winner_break_points_saved', 'winner_break_points_serve_total',
       'winner_service_points_won', 'winner_service_points_total',
       'winner_first_serve_return_won', 'winner_first_serve_return_total',
       'winner_second_serve_return_won', 'winner_second_serve_return_total',
       'winner_break_points_converted', 'winner_break_points_return_total',
       'winner_total_points_won', 'winner_total_points_total'
       'loser_name','loser_first_serves_in','loser_first_serves_total',
       'loser_first_serve_points_won','loser_first_serve_points_total',
       'loser_second_serve_points_won','loser_second_serve_points_total',
        'loser_break_points_saved','loser_break_points_serve_total',
        'loser_service_points_won','loser_service_points_total',
        'loser_first_serve_return_won','loser_first_serve_return_total',
        'loser_second_serve_return_won','loser_second_serve_return_total',
        'loser_break_points_converted', 'loser_break_points_return_total',
        'loser_total_points_won', 'loser_total_points_total']
df = df[[c for c in df.columns if c in lst]]
df.head()

winner_first_serves_in  winner_first_serves_total  \
0                     0.0                       65.0   
1                    44.0                       65.0   
2                    17.0                       36.0   
3                    44.0                       58.0   
4                    37.0                       71.0   

   winner_first_serve_points_won  winner_first_serve_points_total  \
0                           29.0                             46.0   
1                           31.0                             44.0   
2                           11.0                             17.0   
3                           28.0                             44.0   
4                           25.0                             37.0   

   winner_second_serve_points_won  winner_second_serve_points_total  \
0                            12.0                              19.0   
1                            12.0                              21.0   
2                            13.0                              19.0   
3                             7.0                              14.0   
4                            19.0                              34.0   

   winner_break_points_saved  winner_break_points_serve_total  \
0                        0.0                              2.0   
1                        2.0                              5.0   
2                        1.0                              3.0   
3                        6.0                              8.0   
4                        1.0                              3.0   

   winner_service_points_won  winner_service_points_total  ...  \
0                       41.0                         65.0  ...   
1                       43.0                         65.0  ...   
2                       24.0                         36.0  ...   
3                       35.0                         58.0  ...   
4                       44.0                         71.0  ...   

   loser_service_points_total  loser_first_serve_return_won  \
0                        69.0                          17.0   
1                        83.0                          13.0   
2                        54.0                           6.0   
3                        56.0                          16.0   
4                        64.0                          12.0   

   loser_first_serve_return_total  loser_second_serve_return_won  \
0                            46.0                            7.0   
1                            44.0                            9.0   
2                            17.0                            6.0   
3                            44.0                            7.0   
4                            37.0                           15.0   

   loser_second_serve_return_total  loser_break_points_converted  \
0                             19.0                           2.0   
1                             21.0                           3.0   
2                             19.0                           2.0   
3                             14.0                           2.0   
4                             34.0                           2.0   

   loser_break_points_return_total  loser_total_points_won  \
0                              2.0                    61.0   
1                              5.0                    68.0   
2                              3.0                    32.0   
3                              8.0                    52.0   
4                              3.0                    59.0   

   loser_total_points_total    winner_name  
0                     134.0  Nicklas Kulti  
1                     148.0  Nicklas Kulti  
2                      90.0  Nicklas Kulti  
3                     114.0    Jimmy Arias  
4                     135.0  Patrik Kuhnen  

[5 rows x 36 columns]

In [3]:
def read_csv_training():
    df = pd.read_csv("Dades/df_train_complete.csv")
    # df = pd.read_csv("Dades/df_train_complete.csv", skiprows = range(1, 35000), skipfooter = 30000)
    str_to_year = [int(year[0:4]) for year in df['match_id']]
    df['match_id'] = str_to_year
    
#     print(df['winner_double_faults'])
    lst_winner = ['winner_name','winner_first_serves_in', 'winner_first_serves_total',
         'winner_first_serve_points_won', 'winner_first_serve_points_total',
         'winner_second_serve_points_won', 'winner_second_serve_points_total',
       'winner_break_points_saved', 'winner_break_points_serve_total',
       'winner_service_points_won', 'winner_service_points_total',
       'winner_first_serve_return_won', 'winner_first_serve_return_total',
       'winner_second_serve_return_won', 'winner_second_serve_return_total',
       'winner_break_points_converted', 'winner_break_points_return_total','winner_double_faults',
        'winner_total_points_won', 'winner_total_points_total']
    lst_loser = ['loser_name','loser_first_serves_in','loser_first_serves_total',
       'loser_first_serve_points_won','loser_first_serve_points_total',
       'loser_second_serve_points_won','loser_second_serve_points_total',
        'loser_break_points_saved','loser_break_points_serve_total',
        'loser_service_points_won','loser_service_points_total',
        'loser_first_serve_return_won','loser_first_serve_return_total',
        'loser_second_serve_return_won','loser_second_serve_return_total',
        'loser_break_points_converted', 'loser_break_points_return_total','loser_double_faults',
        'loser_total_points_won', 'loser_total_points_total']
    
#     df_winner = df[[c for c in df.columns if c in lst_winner]]
    df_winner = df[lst_winner]
#     df_loser = df[[c for c in df.columns if c in lst_loser]]
    df_loser = df[lst_loser]
    
    df_winner.rename(columns = {'winner_name':'player_name'}, inplace = True)
    df_loser.rename(columns = {'loser_name':'player_name'}, inplace = True)
    
    df_winner = df_winner.groupby('player_name').agg(['sum', 'count'])
#     df_winner = df_winner.reset_index()
    df_loser = df_loser.groupby('player_name').agg(['sum', 'count'])
#     df_loser = df_loser.reset_index()
    
    df = pd.merge(df_winner, df_loser, how ='inner', on = 'player_name')
    
#     TODO: Ajuntar winner i loser data
    lst = ['player_first_serves_in','player_first_serves_total',
       'player_first_serve_points_won','player_first_serve_points_total',
       'player_second_serve_points_won','player_second_serve_points_total',
        'player_break_points_saved','player_break_points_serve_total',
        'player_service_points_won','player_service_points_total',
        'player_first_serve_return_won','player_first_serve_return_total',
        'player_second_serve_return_won','player_second_serve_return_total',
        'player_break_points_converted', 'player_break_points_return_total','player_double_faults',
        'player_total_points_won', 'player_total_points_total']
    for i in range(len(lst)):
        for j in ['sum', 'count']:
            df[lst[i], j] = df[lst_winner[i+1], j] + df[lst_loser[i+1], j]
            
    df = df.iloc[:,len(lst)*4:len(lst)*6]
    
    
    return df

## Simulació test 1 (Màxim variables)

In [4]:
def normalize_probs(prob_1, prob_2):
    total = prob_1 + prob_2
    return prob_1/total, prob_2/total

In [5]:
# Funció simple:
def point_sim(server_pct, receiver_pct):
    server_pct, receiver_pct = normalize_probs(server_pct, receiver_pct)
    x = random.random()
    if x < server_pct:
        return 1
    else:
        return 0

In [6]:
def tiebreak_sim(player_a, player_b):
    points_a = 0
    points_b = 0
    switch = True
    while points_a < 7 and points_b < 7:
        if (points_a + points_b) % 2 == 1:
            switch = not(switch)
        if switch == 0:
            a_pct = player_a['player_break_points_saved', 'sum'].values[0]/player_a['player_break_points_serve_total', 'sum'].values[0]
            b_pct = player_b['player_break_points_converted', 'sum'].values[0]/player_b['player_break_points_return_total', 'sum'].values[0]
            point = point_sim(a_pct,b_pct)
            if point == 1:
                if points_a == 5 and points_b == 6:
                    points_b -= 1
                else: 
                    points_a += 1
            elif point == 0:
                if points_a == 6 and points_b == 5:
                    points_a -= 1
                else: 
                    points_b += 1
#             print('Switch: ', switch, 'Points_A: ', points_a, 'Points B: ', points_b, 'Pooint: ', point)
        elif switch == 1:
            a_pct = player_a['player_break_points_converted', 'sum'].values[0]/player_a['player_break_points_return_total', 'sum'].values[0]
            b_pct = player_b['player_break_points_saved', 'sum'].values[0]/player_b['player_break_points_serve_total', 'sum'].values[0]
            point = point_sim(b_pct, a_pct)
            if point == 1:
                if points_b == 5 and points_a == 6:
                    points_a -= 1
                else:
                    points_b += 1
            elif point == 0:
                if points_b == 6 and points_a == 5:
                    points_b -= 1
                else:
                    points_a += 1
#             print('Switch: ', switch, 'Points_A: ', points_a, 'Points B: ', points_b, 'Pooint: ', point)
            
    if points_a > points_b:
        return 1
    else:
        return 0
    

In [7]:
# https://towardsdatascience.com/building-a-tennis-match-simulator-in-python-3add9af6bebe
# S'utilitza la lògica del sistema de punts d'aquesta pàgina.

def game_sim(server,receiver): 
    server_points = 0
    receiver_points = 0
    server_prob = ((server['player_first_serves_in', 'sum'].values[0] / server['player_first_serves_total', 'sum'].values[0])
    *(server['player_first_serve_points_won', 'sum'].values[0]/ server['player_first_serve_points_total', 'sum'].values[0])
    +(1 - server['player_first_serves_in', 'sum'].values[0] / server['player_first_serves_total', 'sum'].values[0])
    *(1 - server['player_double_faults', 'sum'].values[0]/ server['player_second_serve_points_total','sum'].values[0])
    *(server['player_second_serve_points_won','sum'].values[0]/server['player_second_serve_points_total','sum'].values[0]))
         
    receiver_prob = ((server['player_first_serves_in', 'sum'].values[0] / server['player_first_serves_total', 'sum'].values[0])
    *(receiver['player_first_serve_return_won','sum'].values[0]/ receiver['player_first_serve_return_total','sum'].values[0])
    +(server['player_double_faults', 'sum'].values[0]/ server['player_second_serve_points_total','sum'].values[0])
    +(1 - (server['player_first_serves_in', 'sum'].values[0] / server['player_first_serves_total', 'sum'].values[0]))
    *(receiver['player_second_serve_return_won', 'sum'].values[0]/ receiver['player_second_serve_return_total','sum'].values[0]))
#     server_prob = server
#     receiver_prob = receiver
    
    while server_points < 4 and receiver_points < 4: 
        point = point_sim(server_prob, receiver_prob)
        if point == 1:
            if server_points == 2 and receiver_points == 3:
                receiver_points -= 1
            else: 
                server_points += 1
        else:
            if server_points == 3 and receiver_points == 2:
                server_points -= 1
            else:
                receiver_points += 1
#         print('Server points:', server_points, ' Receiver points:', receiver_points, ' Point:', point)

#         BREAKPOINT SITUATION

        if receiver_points == 3:
            server_break = server['player_break_points_saved', 'sum'].values[0]/server['player_break_points_serve_total', 'sum'].values[0]
            receiver_break = receiver['player_break_points_converted', 'sum'].values[0]/receiver['player_break_points_return_total', 'sum'].values[0]
            point = point_sim(server_break,receiver_break)
#             point = point_sim(server, receiver)
            if point == 1:
                if server_points == 2:
                    receiver_points -= 1
                else: 
                    server_points += 1
            else:
                receiver_points += 1
                
#             print('Server points:', server_points, ' Receiver points:', receiver_points, ' Point:', point)


    if server_points == 4: 
        return 1
    elif receiver_points == 4:
        return 0

In [8]:
def set_sim(player_a, player_b):
    games_a = 0
    games_b = 0
    games = 0
    while games_a < 6 and games_b < 6:
        if games%2 == 0:
#             print('Server player A - Receiver player B')
            game = game_sim(player_a, player_b)
            if game == 1:
                games_a += 1
            else:
                games_b +=1
        else:
#             print('Server player B - Receiver player A')
            game = game_sim(player_b, player_a)
            if game == 1:
                games_b += 1
            else:
                games_a +=1
        games += 1
#         print('Game A:', games_a, '------ Games B:', games_b, ' Game:', game)
    if abs(games_a - games_b) < 2:
        if games%2 == 0:
#             print('Server player A - Receiver player B')
            game = game_sim(player_a, player_b)
            if game == 1:
                games_a += 1
            else:
                games_b +=1
        else:
#             print('Server player B - Receiver player A')
            game = game_sim(player_b, player_a)
            if game == 1:
                games_b += 1
            else:
                games_a +=1
#         print('Game A:', games_a, '------ Games B:', games_b)
        if games_a == games_b:
#             print('tiebreak')
            tiebreak = tiebreak_sim(player_a, player_b)
            if tiebreak == 1:
                return 1
            else:
                return 0
        elif games_a > games_b:
            return 1
        elif games_b > games_a:
            return 0
    else:
        if games_a == 6:
            return 1
        elif games_b == 6:
            return 0

        

In [9]:
def match_sim(player_a, player_b, num_set):
    sets_a = 0
    sets_b = 0 
    sets_played = 0
    while sets_played < num_set and (sets_a < ((num_set//2) +1) or sets_b < ((num_set//2) +1)):
        set_result = set_sim(player_a, player_b)
        if set_result == 1:
            sets_a += 1
        else:
            sets_b += 1
        sets_played += 1
#         print('Sets A: ', sets_a, ' Sets B: ', sets_b, 'Sets Played: ', sets_played)
#         print('---------------------------------------------------------------------')
    if sets_a > sets_b:
        return 1
    elif sets_b > sets_a:
        return 0

In [10]:
# print(match_sim(0.5,0.5,3))

In [11]:
def read_csv_test():
    df_test = pd.read_csv("Dades/df_test_complete.csv")

    str_to_year = [int(year[0:4]) for year in df_test['match_id']]
    df_test['match_id'] = str_to_year
#     print(len(df_test['match_id']))

    return(df_test)

In [12]:
df_test = read_csv_test()
df_final = read_csv_training()
count = 0
count_success = 0
for index, row in df_test.iterrows():
        index_winner = df_final.index[df_final.index == row['winner_name']].tolist()
        index_loser = df_final.index[df_final.index == row['loser_name']].tolist()
        if index_winner and index_loser:
#             print(row['winner_name'], '-------' ,row['loser_name'])
#             f = df_final.loc[index_winner]
#             print(f['player_first_serves_in', 'sum'].values[0])
            result = [match_sim(df_final.loc[index_winner], df_final.loc[index_loser], 3) for i in range(11)]
            count += 1
            if sum(result) >= 6:
                count_success += 1
print('Count Success: ', count_success/count)
print(count_success)
print(count)

C:\Users\Usuari\AppData\Local\Temp/ipykernel_2196/2307590342.py:54: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead.  To get a de-fragmented frame, use `newframe = frame.copy()`
  df[lst[i], j] = df[lst_winner[i+1], j] + df[lst_loser[i+1], j]


Count Success:  0.5547030406365445
1952
3519


## Simulació Simple:

In [13]:
def normalize_probs(prob_1, prob_2):
    total = prob_1 + prob_2
    return prob_1/total, prob_2/total

In [14]:
# Funció simple:
def point_sim(server_pct, receiver_pct):
    server_pct, receiver_pct = normalize_probs(server_pct, receiver_pct)
    x = random.random()
    if x < server_pct:
        return 1
    else:
        return 0

In [15]:
def tiebreak_sim(player_a, player_b):
    points_a = 0
    points_b = 0
    switch = True
    a_pct = ((player_a['player_total_points_won', 'sum'].values[0] / player_a['player_total_points_total', 'sum'].values[0]))
    b_pct = ((player_b['player_total_points_won', 'sum'].values[0] / player_b['player_total_points_total', 'sum'].values[0]))
    while points_a < 7 and points_b < 7:
        if (points_a + points_b) % 2 == 1:
            switch = not(switch)
        if switch == 0:
            point = point_sim(a_pct,b_pct)
            if point == 1:
                if points_a == 5 and points_b == 6:
                    points_b -= 1
                else: 
                    points_a += 1
            elif point == 0:
                if points_a == 6 and points_b == 5:
                    points_a -= 1
                else: 
                    points_b += 1
#             print('Switch: ', switch, 'Points_A: ', points_a, 'Points B: ', points_b, 'Pooint: ', point)
        elif switch == 1:
            point = point_sim(b_pct, a_pct)
            if point == 1:
                if points_b == 5 and points_a == 6:
                    points_a -= 1
                else:
                    points_b += 1
            elif point == 0:
                if points_b == 6 and points_a == 5:
                    points_b -= 1
                else:
                    points_a += 1
#             print('Switch: ', switch, 'Points_A: ', points_a, 'Points B: ', points_b, 'Pooint: ', point)
            
    if points_a > points_b:
        return 1
    else:
        return 0

In [16]:
def game_sim(server,receiver): 
    server_points = 0
    receiver_points = 0
    server_prob = ((server['player_total_points_won', 'sum'].values[0] / server['player_total_points_total', 'sum'].values[0]))
    receiver_prob = ((receiver['player_total_points_won', 'sum'].values[0] / receiver['player_total_points_total', 'sum'].values[0]))
    
    while server_points < 4 and receiver_points < 4: 
        point = point_sim(server_prob, receiver_prob)
        if point == 1:
            if server_points == 2 and receiver_points == 3:
                receiver_points -= 1
            else: 
                server_points += 1
        else:
            if server_points == 3 and receiver_points == 2:
                server_points -= 1
            else:
                receiver_points += 1
#         print('Server points:', server_points, ' Receiver points:', receiver_points, ' Point:', point)

#         BREAKPOINT SITUATION

        if receiver_points == 3:
            point = point_sim(server_prob,receiver_prob)
#             point = point_sim(server, receiver)
            if point == 1:
                if server_points == 2:
                    receiver_points -= 1
                else: 
                    server_points += 1
            else:
                receiver_points += 1
                
#             print('Server points:', server_points, ' Receiver points:', receiver_points, ' Point:', point)


    if server_points == 4: 
        return 1
    elif receiver_points == 4:
        return 0

In [17]:
def set_sim(player_a, player_b):
    games_a = 0
    games_b = 0
    games = 0
    while games_a < 6 and games_b < 6:
        if games%2 == 0:
#             print('Server player A - Receiver player B')
            game = game_sim(player_a, player_b)
            if game == 1:
                games_a += 1
            else:
                games_b +=1
        else:
#             print('Server player B - Receiver player A')
            game = game_sim(player_b, player_a)
            if game == 1:
                games_b += 1
            else:
                games_a +=1
        games += 1
#         print('Game A:', games_a, '------ Games B:', games_b, ' Game:', game)
    if abs(games_a - games_b) < 2:
        if games%2 == 0:
#             print('Server player A - Receiver player B')
            game = game_sim(player_a, player_b)
            if game == 1:
                games_a += 1
            else:
                games_b +=1
        else:
#             print('Server player B - Receiver player A')
            game = game_sim(player_b, player_a)
            if game == 1:
                games_b += 1
            else:
                games_a +=1
#         print('Game A:', games_a, '------ Games B:', games_b)
        if games_a == games_b:
#             print('tiebreak')
            tiebreak = tiebreak_sim(player_a, player_b)
            if tiebreak == 1:
                return 1
            else:
                return 0
        elif games_a > games_b:
            return 1
        elif games_b > games_a:
            return 0
    else:
        if games_a == 6:
            return 1
        elif games_b == 6:
            return 0


In [18]:
def match_sim(player_a, player_b, num_set):
    sets_a = 0
    sets_b = 0 
    sets_played = 0
    while sets_played < num_set and (sets_a < ((num_set//2) +1) or sets_b < ((num_set//2) +1)):
        set_result = set_sim(player_a, player_b)
        if set_result == 1:
            sets_a += 1
        else:
            sets_b += 1
        sets_played += 1
#         print('Sets A: ', sets_a, ' Sets B: ', sets_b, 'Sets Played: ', sets_played)
#         print('---------------------------------------------------------------------')
    if sets_a > sets_b:
        return 1
    elif sets_b > sets_a:
        return 0

In [19]:
def read_csv_test():
    df_test = pd.read_csv("Dades/df_test_complete.csv")

    str_to_year = [int(year[0:4]) for year in df_test['match_id']]
    df_test['match_id'] = str_to_year
#     print(len(df_test['match_id']))

    return(df_test)

In [20]:
df_test = read_csv_test()
df_final = read_csv_training()
count = 0
count_success = 0
for index, row in df_test.iterrows():
        index_winner = df_final.index[df_final.index == row['winner_name']].tolist()
        index_loser = df_final.index[df_final.index == row['loser_name']].tolist()
        if index_winner and index_loser:
#             print(row['winner_name'], '-------' ,row['loser_name'])
#             f = df_final.loc[index_winner]
#             print(f['player_first_serves_in', 'sum'].values[0])
            result = [match_sim(df_final.loc[index_winner], df_final.loc[index_loser], 3) for i in range(11)]
            count += 1
            if sum(result) >= 6:
                count_success += 1
print('Count Success: ', count_success/count)
print(count_success)
print(count)

C:\Users\Usuari\AppData\Local\Temp/ipykernel_2196/2307590342.py:54: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead.  To get a de-fragmented frame, use `newframe = frame.copy()`
  df[lst[i], j] = df[lst_winner[i+1], j] + df[lst_loser[i+1], j]


Count Success:  0.545609548167093
1920
3519


## Simulació %Victòries

In [21]:
def read_csv_training():
    df = pd.read_csv("Dades/df_train_complete.csv")
    # df = pd.read_csv("Dades/df_train_complete.csv", skiprows = range(1, 35000), skipfooter = 30000)
    str_to_year = [int(year[0:4]) for year in df['match_id']]
    df['match_id'] = str_to_year
    
#     print(df['winner_double_faults'])
    lst_winner = ['winner_name','winner_first_serves_in']
    lst_loser = ['loser_name','loser_first_serves_in']
    
#     df_winner = df[[c for c in df.columns if c in lst_winner]]
    df_winner = df[lst_winner]
#     df_loser = df[[c for c in df.columns if c in lst_loser]]
    df_loser = df[lst_loser]
    
    df_winner.rename(columns = {'winner_name':'player_name', 'winner_first_serves_in': 'win_count'}, inplace = True)
    df_loser.rename(columns = {'loser_name':'player_name', 'loser_first_serves_in': 'loose_count'}, inplace = True)
    
    df_winner = df_winner.groupby('player_name').count()
#     df_winner = df_winner.reset_index()
    df_loser = df_loser.groupby('player_name').count()
#     df_loser = df_loser.reset_index()
    
    df = pd.merge(df_winner, df_loser, how ='inner', on = 'player_name')
    
    df['win_pct'] = df['win_count']/(df['win_count'] + df['loose_count'])
    
    return df

In [22]:
def normalize_probs(prob_1, prob_2):
    total = prob_1 + prob_2
    return prob_1/total, prob_2/total

In [23]:
# Funció simple:
def point_sim(server_pct, receiver_pct):
    server_pct, receiver_pct = normalize_probs(server_pct, receiver_pct)
    x = random.random()
    if x < server_pct:
        return 1
    else:
        return 0

In [24]:
def tiebreak_sim(player_a, player_b):
    points_a = 0
    points_b = 0
    switch = True
    a_pct = player_a['win_pct'].values[0] 
    b_pct = player_b['win_pct'].values[0]
    while points_a < 7 and points_b < 7:
        if (points_a + points_b) % 2 == 1:
            switch = not(switch)
        if switch == 0:
            point = point_sim(a_pct,b_pct)
            if point == 1:
                if points_a == 5 and points_b == 6:
                    points_b -= 1
                else: 
                    points_a += 1
            elif point == 0:
                if points_a == 6 and points_b == 5:
                    points_a -= 1
                else: 
                    points_b += 1
#             print('Switch: ', switch, 'Points_A: ', points_a, 'Points B: ', points_b, 'Pooint: ', point)
        elif switch == 1:
            point = point_sim(b_pct, a_pct)
            if point == 1:
                if points_b == 5 and points_a == 6:
                    points_a -= 1
                else:
                    points_b += 1
            elif point == 0:
                if points_b == 6 and points_a == 5:
                    points_b -= 1
                else:
                    points_a += 1
#             print('Switch: ', switch, 'Points_A: ', points_a, 'Points B: ', points_b, 'Pooint: ', point)
            
    if points_a > points_b:
        return 1
    else:
        return 0

In [25]:
def game_sim(server,receiver): 
    server_points = 0
    receiver_points = 0
    server_prob = server['win_pct'].values[0] 
    receiver_prob = receiver['win_pct'].values[0]
    
    while server_points < 4 and receiver_points < 4: 
        point = point_sim(server_prob, receiver_prob)
        if point == 1:
            if server_points == 2 and receiver_points == 3:
                receiver_points -= 1
            else: 
                server_points += 1
        else:
            if server_points == 3 and receiver_points == 2:
                server_points -= 1
            else:
                receiver_points += 1
#         print('Server points:', server_points, ' Receiver points:', receiver_points, ' Point:', point)

#         BREAKPOINT SITUATION

        if receiver_points == 3:
            point = point_sim(server_prob,receiver_prob)
#             point = point_sim(server, receiver)
            if point == 1:
                if server_points == 2:
                    receiver_points -= 1
                else: 
                    server_points += 1
            else:
                receiver_points += 1
                
#             print('Server points:', server_points, ' Receiver points:', receiver_points, ' Point:', point)


    if server_points == 4: 
        return 1
    elif receiver_points == 4:
        return 0

In [26]:
def set_sim(player_a, player_b):
    games_a = 0
    games_b = 0
    games = 0
    while games_a < 6 and games_b < 6:
        if games%2 == 0:
#             print('Server player A - Receiver player B')
            game = game_sim(player_a, player_b)
            if game == 1:
                games_a += 1
            else:
                games_b +=1
        else:
#             print('Server player B - Receiver player A')
            game = game_sim(player_b, player_a)
            if game == 1:
                games_b += 1
            else:
                games_a +=1
        games += 1
#         print('Game A:', games_a, '------ Games B:', games_b, ' Game:', game)
    if abs(games_a - games_b) < 2:
        if games%2 == 0:
#             print('Server player A - Receiver player B')
            game = game_sim(player_a, player_b)
            if game == 1:
                games_a += 1
            else:
                games_b +=1
        else:
#             print('Server player B - Receiver player A')
            game = game_sim(player_b, player_a)
            if game == 1:
                games_b += 1
            else:
                games_a +=1
#         print('Game A:', games_a, '------ Games B:', games_b)
        if games_a == games_b:
#             print('tiebreak')
            tiebreak = tiebreak_sim(player_a, player_b)
            if tiebreak == 1:
                return 1
            else:
                return 0
        elif games_a > games_b:
            return 1
        elif games_b > games_a:
            return 0
    else:
        if games_a == 6:
            return 1
        elif games_b == 6:
            return 0


In [27]:
def match_sim(player_a, player_b, num_set):
    sets_a = 0
    sets_b = 0 
    sets_played = 0
    while sets_played < num_set and (sets_a < ((num_set//2) +1) or sets_b < ((num_set//2) +1)):
        set_result = set_sim(player_a, player_b)
        if set_result == 1:
            sets_a += 1
        else:
            sets_b += 1
        sets_played += 1
#         print('Sets A: ', sets_a, ' Sets B: ', sets_b, 'Sets Played: ', sets_played)
#         print('---------------------------------------------------------------------')
    if sets_a > sets_b:
        return 1
    elif sets_b > sets_a:
        return 0

In [28]:
def read_csv_test():
    df_test = pd.read_csv("Dades/df_test_complete.csv")

    str_to_year = [int(year[0:4]) for year in df_test['match_id']]
    df_test['match_id'] = str_to_year
#     print(len(df_test['match_id']))

    return(df_test)

In [29]:
df_test = read_csv_test()
df_final = read_csv_training()
count = 0
count_success = 0
for index, row in df_test.iterrows():
        index_winner = df_final.index[df_final.index == row['winner_name']].tolist()
        index_loser = df_final.index[df_final.index == row['loser_name']].tolist()
        if index_winner and index_loser:
#             print(row['winner_name'], '-------' ,row['loser_name'])
#             f = df_final.loc[index_winner]
#             print(f['player_first_serves_in', 'sum'].values[0])
            result = [match_sim(df_final.loc[index_winner], df_final.loc[index_loser], 3) for i in range(11)]
            count += 1
            if sum(result) >= 6:
                count_success += 1
print('Count Success: ', count_success/count)
print(count_success)
print(count)

Count Success:  0.6126740551292981
2156
3519


## Resultats:

Simulació al màxim: **0.55470**  
Simulació punts_pct: **0.54561**  
Simulació win_pct: **0.61267**